integration/process_execution_heartbeat_test.go

```diff
@@ -0,0 +1,154 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// +build integration
+
+package integration
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3em/agent"
+	"github.com/m3db/m3em/build"
+	hb "github.com/m3db/m3em/generated/proto/heartbeat"
+	"github.com/m3db/m3em/generated/proto/m3em"
+	"github.com/m3db/m3em/operator"
+
+	"github.com/m3db/m3x/instrument"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+)
+
+func TestProcessExecutionHeartbeating(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+	targetLocation := newTempDir(t)
+	defer os.RemoveAll(targetLocation)
+
+	iopts := instrument.NewOptions()
+	// create agent listener, get free port from OS
+	agentListener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer agentListener.Close()
+
+	// create heartbeat listener, get free port from OS
+	heartbeatListener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer heartbeatListener.Close()
+
+	// create agent
+	aOpts := agent.NewOptions(iopts).SetWorkingDirectory(targetLocation)
+	server := grpc.NewServer(grpc.MaxConcurrentStreams(16384))
+	agentService, err := agent.New(aOpts)
+	require.NoError(t, err)
+	m3em.RegisterOperatorServer(server, agentService)
+	go func() {
+		server.Serve(agentListener)
+	}()
+	defer server.GracefulStop()
+
+	// create new heartbeat router
+	hbRouter := operator.NewHeartbeatRouter(heartbeatListener.Addr().String())
+	hbServer := grpc.NewServer(grpc.MaxConcurrentStreams(16384))
+	hb.RegisterHeartbeaterServer(hbServer, hbRouter)
+	go func() {
+		hbServer.Serve(heartbeatListener)
+	}()
+	defer hbServer.GracefulStop()
+
+	// create operator to communicate with agent
+	hbOpts := operator.NewHeartbeatOptions().
+		SetEnabled(true).
+		SetHeartbeatRouter(hbRouter)
+	oOpts := operator.NewOptions(iopts).
+		SetHeartbeatOptions(hbOpts)
+	op, err := operator.New(agentListener.Addr().String(), oOpts)
+	require.NoError(t, err)
+
+	// capture notifications from operator
+	var (
+		lock                sync.Mutex
+		notifiedTermination = false
+	)
+	operatorEventListener := operator.NewListener(
+		func(s string) {
+			require.False(t, notifiedTermination)
+			lock.Lock()
+			notifiedTermination = true
+			lock.Unlock()
+		},
+		func(ts time.Time) {
+			require.FailNow(t, "received timeout: %v", ts.String())
+		},
+		func(s string) {
+			require.FailNow(t, "received overwrite: %v", s)
+		},
+	)
+	id := op.RegisterListener(operatorEventListener)
+	defer op.DeregisterListener(id)
+
+	// create test build
+	scriptContents := []byte(`#!/usr/bin/env bash
```

Comment:
maybe we can make a `testProcessExecutionHeartbeating.sh` file and read it in instead of inlining bash?